narwhal/primary/src/block_remover.rs

```diff
@@ -209,190 +49,78 @@ pub struct BlockRemover {
-                        for batch_ids in batches_by_worker.values() {
-                            let request_key = Self::construct_batches_request_key(batch_ids);
-
-                            self.map_tx_worker_removal_results.remove(&request_key);
-                        }
+    pub async fn remove_blocks(&self, ids: Vec<CertificateDigest>) -> Result<()> {
+        // Look up certificates requested for removal.
+        let certificates = self.certificate_store.read_all(ids.clone())?;
+        let non_found_digests: Vec<CertificateDigest> = certificates
+            .iter()
+            .zip(ids.clone())
+            .filter_map(|(c, digest)| if c.is_none() { Some(digest) } else { None })
+            .collect();
+        if !non_found_digests.is_empty() {
+            warn!("Some certificates are missing, will ignore for removal: {non_found_digests:?}");
+        }
+        let found_certificates: Vec<Certificate> = certificates.into_iter().flatten().collect();
 
-                        // Do the further clean up only if the result is successful
-                        if result.is_ok() {
-                            self.cleanup_internal_state(certificates, batches_by_worker)
-                                .await
-                                .map_err(|_err| ())
-                        } else {
-                            Ok(())
-                        }
-                    }
-                };
+        // Send delete requests to individual workers.
+        let mut worker_requests = Vec::new();
+        let batches_by_worker =
+            utils::map_certificate_batches_by_worker(found_certificates.as_slice());
+        for (worker_id, batch_ids) in batches_by_worker.iter() {
+            let worker_name = self
+                .worker_cache
+                .load()
+                .worker(&self.name, worker_id)
+                .expect("Worker id not found")
+                .name;
 
-                // if the clean up is successful we want to send the result
-                // whatever this is. Otherwise, we are sending back an error
-                // response.
-                if cleanup_successful.is_ok() {
-                    Self::unreliable_broadcast(senders, result).await;
-                } else {
-                    Self::unreliable_broadcast(
-                        senders,
-                        Err(BlockRemoverError {
-                            ids: block_ids,
-                            error: BlockRemoverErrorKind::Failed,
-                        }),
-                    )
-                    .await;
-                }
-            }
+            debug!(
+                "Sending DeleteBatches request for batch ids {batch_ids:?} to worker {worker_name}"
+            );
+            worker_requests.push(
+                self.worker_network
+                    .delete_batches(worker_name, batch_ids.clone()),
+            );
         }
-    }
+        try_join_all(worker_requests).await?;
```

Comment:
`try_join_all` fails fast if any future returns error, without waiting for everything else to complete - is this the desired behavior here?

Reply:
(nvm, this actually seem correct)

Reply:
Yes it is, this matches the previous behavior